src/plugins/intel_cpu/src/infer_request.cpp

```diff
@@ -47,19 +47,20 @@ void InferRequestBase::CreateInferRequest() {
     // producer as storage for tensor to keep it between infer calls.
     for (auto& node : graph->GetNodes()) {
         if (node->getType() == Type::MemoryInput) {
-            auto memoryNode = dynamic_cast<node::MemoryInput*>(node.get());
+            auto memoryNode = std::dynamic_pointer_cast<node::MemoryInput>(node);
             if (!memoryNode) {
                 IE_THROW() << "Cannot cast " << node->getName() << " to MemoryInput";
             }
-            auto state_store = memoryNode->getStore();
             auto state_name = memoryNode->getId();
 
             // Remove suffix with pair ID. Internal information.
             auto suffix_idx = state_name.find("/id=");
-            if (suffix_idx != std::string::npos)
+            if (suffix_idx != std::string::npos) {
                 state_name = state_name.substr(0, suffix_idx);
+            }
 
-            memoryStates.emplace_back(new VariableState(state_name, state_store));
+            memoryStates.emplace_back(
+                std::make_shared<VariableStateDoubleBuffer>(state_name, memoryNode->memoryBuilder(), memoryNode->getMemoryPtr()));
```

Comment:
It might make sense to move the instantiation of the state objects to the MemoryInput nodes. That will allow to hide the selection of the concrete state type on the graph/node level.

Reply:
done